### ISSUE TITLE
environment variable name transformation question

### ISSUE THREAD

**Issue by jkburges** [buildkite/agent#2100]
We are using the zacharymctague/aws-ssm plugin to populate env vars from AWS SSM, pipeline.yml snippet:

```
    plugins:
      - zacharymctague/aws-ssm#v1.0.0:
          parameters:
            BUNDLE_GEMS__CONTRIBSYS__COM: /ci/sidekiq_pro_api_key
```

Now the sidekiq gem, for which we are configuring the env var, expects it to be as written - i.e. with the double underscores - but I think the Buildkite agent code is transforming it to in env var BUNDLE_GEMS_CONTRIBSYS_COM (single underscores), which is not what we want.

We can probably work around this but I was wondering why this manipulation (replacing multiple underscores with single) is happening?

Code which currently does it I think: https://github.com/buildkite/agent/blob/main/agent/plugin/plugin.go#L211, but the logic seems to have been introduced originally at https://github.com/buildkite/agent/commit/acf4f7d16ef6d5e0afec952b098790d396156b6d#diff-ed8ae83218edc2884a09b5742c7c139b1a1340b38ddb31045bc42507e70c9fb7R177.

**Comment by triarius** [comment]
Hi @jkburges, thanks for reporting this!

You can probably tell from the code that there is a series of transformation made to turn yaml keys into environment variables. Some of these are to replace characters that can't appear in environment variable names, though it is by no means comprehensive. The reasoning for replacing consecutive underscores with a single underscore is not as clear, but I don't think it is out of any necessity, so not applying this transformation is an option.

I've discussed this with the team, and we've decided that to maintain compatibility, we think the best way to this is to (at least for a period) export both versions of the variable name - one allowing consecutive underscores, and only not. Then, the SSM plugin should be able to fill in both with the same value from the SSM parameter store.

In the meantime, I've [forked](https://github.com/triarius/aws-ssm-buildkite-plugin) our version of the plugin, https://github.com/buildkite-plugins/aws-ssm-buildkite-plugin, to use the map of environment variable names to parameter names from the JSON exported as `BUILDKITE_PLUGIN_CONFIGURATION`. This means the `jq` is a dependency of the plugin, so it may not be something we can upstream easily. However, it is available by default on the Elastic CI Stack.

You should be able to use it with something like:
```yaml
    plugins:
      - triarius/aws-ssm#050f04af5f932c5beb4e5b68b214b894fce0ba64:
          parameters:
            BUNDLE_GEMS__CONTRIBSYS__COM: /ci/sidekiq_pro_api_key
```

**Comment by jkburges** [comment]
Thanks @triarius for the info. Is it likely that changes to the agent to include both versions of variable name will make it in to a new version of the agent soon, or is it more likely to be a month or more away? Just deciding whether it's better for us to use the forked plugin, or stick with the workaround we already have and wait for a new agent.

Thanks again 🙏

**Comment by triarius** [comment]
Thanks, @jkburges, the work has started, but it will have to wait for the at least next cycle. 

You may get a notification from GitHub when this issue is closed, but that only means it's been merged into the main branch, not released. As we need to be a bit careful to make sure plugins are not broken, there may be some more testing we need to do while it's only merged and not released.

So I recommend you see if the forked plugin works around the issue for you to unblock yourself now.
